### PR TITLE
CI: Azure - multiple configs

### LINKED ISSUE #12955: Add testing with PyTorch 1.11 on GPUs in CI
## 🚀 Feature
We've decided to have testing with both PyTorch LTS and stable release (1.8 and 1.11 as of now) in CI, and we've already seen some issues arose while trying to enable it in #12373.

### TODO
Known issues with PL with PyTorch 1.11

- [x] #12846
- [x] #12860
- [ ] Fix an issue with fitting a model initialised in init_meta_context #12870
- [x] Fix an issue with DDP comm tests with some newer PyTorch versions #12878
- [ ] Fix an issue with inference mode with FSDP

### Motivation
To test new features, e.g. meta init and native FSDP, in CI that are only available in newer PyTorch versions.

### Pitch
Use the following image:
```
pytorchlightning/pytorch_lightning:base-cuda-py3.7-torch1.11
```

### Alternatives
n/a

### Additional context
n/a

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @carmocca @akihironitta @borda

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Followup #12373
Extend testing configs for GPU also for the latest version
cc: @akihironitta 

Closes #12955.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @carmocca @akihironitta @borda

### CODE REVIEW

**Comment by Borda** [comment]
it is expected to fail the PT1.11 until docker is unpin in #12373

File: .azure-pipelines/gpu-tests.yml [hunk 1]
```diff
@@ -18,6 +18,14 @@ pr:
...
 jobs:
   - job: pytest
+    strategy:
+      matrix:
+        'PyTorch - LTS':
+          pyVersion: '3.7'
+          ptVersion: '1.8'
+        'PyTorch - stable':
+          pyVersion: '3.9'
+          ptVersion: '1.11'
```
Comment (akihironitta) [inline .azure-pipelines/gpu-tests.yml:28]: Sometimes (e.g. when docker images are somehow broken or when unblocking master isn't trivial and needs time), we might want to pin docker images by specifying the hash value, so we could maybe directly specify them here instead of specifying them via `pyVersion` and `ptVersion`.

```suggestion
      matrix:
        'PyTorch - LTS':
          image: "pytorchlightning/pytorch_lightning:base-cuda-py3.7-torch1.8"
        'PyTorch - stable':
          image: "pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.11"
```

**Review by akihironitta — COMMENTED** [review]
🚀

**Review by carmocca — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

**Review by otaj — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

**Review by akihironitta — CHANGES_REQUESTED** [review]
Let me "request changes" to avoid accidentally merging this PR to master.

File: .azure-pipelines/gpu-tests.yml [hunk 2]
```diff
@@ -18,6 +18,12 @@ pr:
...
 jobs:
   - job: pytest
+    strategy:
+      matrix:
+        'PyTorch - LTS':
+          image: "pytorchlightning/pytorch_lightning:base-cuda-py3.7-torch1.8"
+        'PyTorch - stable':
+          image: "pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.11"
     # how long to run the job before automatically cancelling
     timeoutInMinutes: "55"
```
Comment (akihironitta) [inline .azure-pipelines/gpu-tests.yml:28]:
```suggestion
        'PyTorch - LTS':
          image: "pytorchlightning/pytorch_lightning:base-cuda-py3.7-torch1.8"
          timeoutInMinutes: "55"
        'PyTorch - stable':
          image: "pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.11"
          timeoutInMinutes: "65"
    # how long to run the job before automatically cancelling
    timeoutInMinutes: $(timeoutInMinutes)
```

File: .azure-pipelines/gpu-tests.yml [hunk 3]
```diff
@@ -18,18 +18,24 @@ pr:
...
+          timeoutInMinutes: "65"
     # how long to run the job before automatically cancelling
-    timeoutInMinutes: "55"
+    timeoutInMinutes: $(timeoutInMinutes)
     # how much time to give 'run always even if cancelled tasks' before stopping them
     cancelTimeoutInMinutes: "2"
 
     pool: azure-gpus-spot
 
     # ToDo: this need to have installed docker in the base image...
```
Comment (carmocca) [inline .azure-pipelines/gpu-tests.yml:36]: This TODO can also be removed?

Comment (Borda) [reply .azure-pipelines/gpu-tests.yml]: Yep

**Comment by akihironitta** [comment]
For the note, the current failures are:
```
6685:2022-05-10T00:49:27.6990597Z FAILED tests/strategies/test_ddp_fully_sharded_native.py::test_fully_sharded_native_strategy_sync_batchnorm
6797:2022-05-10T00:49:38.9204765Z FAILED tests/strategies/test_ddp_fully_sharded_native.py::test_fully_sharded_native_strategy_checkpoint
6911:2022-05-10T00:49:50.1150632Z FAILED tests/strategies/test_ddp_fully_sharded_native.py::test_fully_sharded_native_strategy_checkpoint_multi_gpus
10517:2022-05-10T00:59:43.0355034Z FAILED tests/strategies/test_deepspeed_strategy.py::test_deepspeed_with_meta_device
11307:2022-05-10T00:59:43.5558192Z FAILED tests/strategies/test_deepspeed_strategy.py::test_deepspeed_with_meta_device
```

File: pytorch_lightning/utilities/imports.py [hunk 4]
```diff
@@ -94,6 +94,8 @@ def _compare_version(package: str, op: Callable, version: str, use_base_version:
 _TORCH_GREATER_EQUAL_1_10 = _compare_version("torch", operator.ge, "1.10.0")
 _TORCH_LESSER_EQUAL_1_10_2 = _compare_version("torch", operator.le, "1.10.2")
 _TORCH_GREATER_EQUAL_1_11 = _compare_version("torch", operator.ge, "1.11.0")
+# todo: remove "dev" when PyTorch 1.12 is released
+_TORCH_GREATER_EQUAL_1_12 = _compare_version("torch", operator.ge, "1.12.0dev")
```
Comment (carmocca) [inline pytorch_lightning/utilities/imports.py:98]:
```suggestion
_TORCH_GREATER_EQUAL_1_12 = _compare_version("torch", operator.ge, "1.12.0" use_base_version=True)
```

**Comment by akihironitta** [comment]
This PR should finally be ready for merge after #12870 lands.

File: tests/standalone_tests.sh [hunk 5]
```diff
@@ -12,7 +12,7 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-set -e
+# set -e  # FIXME
```
Comment (Borda) [inline tests/standalone_tests.sh:15]:
```suggestion
set -e  # FIXME
```

**Review by akihironitta — APPROVED** [review]

**Comment by akihironitta** [comment]
The blocker #12870 has been resolved, so this PR should be ready for getting merged.

@Borda @carmocca Could you have another look if needed?

**Review by carmocca — APPROVED** [review]

File: .azure-pipelines/gpu-tests.yml [hunk 6]
```diff
@@ -18,18 +18,21 @@ pr:
 
 jobs:
   - job: pytest
+    strategy:
+      matrix:
+        'PyTorch - LTS':
+          image: "pytorchlightning/pytorch_lightning:base-cuda-py3.7-torch1.8"
+        'PyTorch - stable':
```
Comment (carmocca) [inline .azure-pipelines/gpu-tests.yml:25]: nit: can we shorten this text?

```suggestion
        'LTS':
          image: "pytorchlightning/pytorch_lightning:base-cuda-py3.7-torch1.8"
        'stable':
```
